apps/labs/posts/scipy-array-api.md

```diff
@@ -0,0 +1,441 @@
+the array API standard.
+
+---
+
+## SciPy and the Scientific Python Ecosystem
+
+SciPy is one part of a wide collection of Open Source tools, written in Python,
+which are fundamental to a range of work which relies on manipulating numbers with computers.
+This collection is known as the Scientific Python Ecosystem - a term which broadly
+encompasses Open Source Python projects which are community driven and used for science
+(see [scientific-python.org](https://scientific-python.org/about/) for more information).
+
+NumPy is the n-dimensional array library which SciPy is built to work with - it is SciPy's only non-optional
+runtime dependency.
+At the time of writing, the vast majority of SciPy functions only work when providing NumPy arrays,
+despite the various interoperability efforts so far
+(see [Anirudh's blog post](https://labs.quansight.org/blog/array-libraries-interoperability)).
+The main problem with this is that "NumPy provides an array implementation that’s in-memory,
+CPU-only and single-threaded."
+The potential performance improvements that could be won with parallel algorithms, support for distributed
+arrays and support for GPUs (and other hardware accelerators) are huge, and thus "SciPy has had 'support
+for distributed and GPU arrays' on its roadmap" for five years now.
+See [this page](https://data-apis.org/array-api/latest/use_cases.html#use-case-scipy) on the use cases of the
+array API and [this blog post](https://labs.quansight.org/blog/2021/11/pydata-extensibility-vision) by
+Ivan Yashchuk and Ralf Gommers for more information.
+
+Due to these performance wishes, many different "array provider" libraries, such as CuPy and
+PyTorch, have been developed, with support for one or more of these hardware use-cases in mind.
+However, due to the fact that "array consumer" libraries are typically built to be compatible with
+only one array provider library, many functions have been reimplemented in different consumer libraries,
+just to work with a different provider library.
+In extreme cases, this has led to efforts to reimplement entire consumer libraries,
+for example [`cupyx.scipy`](https://docs.cupy.dev/en/stable/reference/scipy.html).
+This constant reimplementation of the same functionality has led many to see the Scientific Python Ecosystem
+as rather being fractured into many smaller ecosystems, one built around each array library, and to dream for a
+world where the ecosystems are unified and everything works together!
+
+While it would certainly be a very ambitious goal to aim for _everything_ working together, the goal
+of increasing interoperability between array libraries is one with widespread support.
+There are currently many hurdles which users can run into when navigating this fractured ecosystem.
+They can find that they have to convert between different array provider libraries in their projects, in order to
+make use of the full range of tools which have been developed.
```

Comment:
You could consider adding a line here saying explicitly that such conversation between array libraries is a show-stopper for many projects. Am I going to convert between NumPy and CuPy arrays often? Nope, too much of a difference hit to do CPU-GPU copying all the time. Am I going to convert between NumPy and PyTorch often? Nope, that kills autodifferentiation features (and also GPU perfomance hit like CuPy).

Reply:
```suggestion
make use of the full range of tools which have been developed.
For many projects, frequent conversion between provider libraries is a complete no-go: cross-device copying
(e.g. from GPU to CPU when converting from CuPy to NumPy) can be a huge performance hit, and 
[automatic differentation](https://en.wikipedia.org/wiki/Automatic_differentiation) features are also killed.
```